vissl/trainer/train_task.py

```diff
@@ -696,7 +703,37 @@ def prepare(self, pin_memory: bool = False):
                 self.base_loss.load_state_dict(self.checkpoint["loss"])
                 logging.info("======Loaded loss state from checkpoint======")
 
-        return self._update_classy_state(vissl_state_dict)
+        current_phase_id = (
+            vissl_state_dict["phase_idx"] + 1 if self.train and vissl_state_dict else 0
+        )  # Dataloader expects 0 phase_idx when in test mode.
+        self.dataloaders = self.build_dataloaders(
+            pin_memory=pin_memory, current_phase_id=current_phase_id
+        )
+        task = self._update_classy_state(vissl_state_dict)
+
+        return task
+
+    def set_epoch(self, phase_type: str, epoch: int, start_iter: int):
+        if hasattr(self.dataloaders[phase_type], "sampler"):
+            sampler = self.dataloaders[phase_type].sampler
+            # (Re-)Shuffle data: set epoch of distributed (or fairstore) sampler.
```

Comment:
nit: the following lines could be factorized with below

```
if hasattr(sampler, "set_epoch"):
    sampler.set_epoch(epoch)
# Resume from the iteration if valid
if hasattr(sampler, "set_start_iter"):
    sampler.set_start_iter(start_iter)
```

Another idea would be to introduce a "tagging interface": instead of checking the presence of method, do a `isinstance` check of an interface, that contains the functions you want. The main advantage is that the name of the interface might bring some more additional information.

Reply:
Yep indeed, good catch (this code is existing, just moved).

IMO, I am not loving this `has_attr` pattern in general. I would prefer an object oriented interface w/ default no-op methods that can be overwritten.